cirq/study/flatten_expressions.py

```diff
@@ -0,0 +1,368 @@
+
+    @staticmethod
+    def default_get_param_name(val: sympy.Basic) -> str:
+        if isinstance(val, sympy.Symbol):
+            return val.name
+        return '<{!s}>'.format(val)
+
+    def _next_symbol(self, val: sympy.Basic) -> sympy.Symbol:
+        name = self.get_param_name(val)
+        symbol = sympy.Symbol(name)
+        # Ensure the symbol hasn't already been used
+        collision = 0
+        while symbol in self._taken_symbols:
+            collision += 1
+            symbol = sympy.Symbol('{}_{}'.format(name, collision))
+        return symbol
+
+    def value_of(self, value: Union[sympy.Basic, float, str]
+                ) -> Union[sympy.Basic, float]:
+        """Resolves a symbol or expression to a new symbol unique to that value.
+
+        If value is a float, returns it.
+        If value is a str, treat it as a symbol with that name and continue.
+        If this `_ParamFlattener` was initialized with a `param_dict` and
+        `value` is a key, returns `param_dict[value]`.
+        Otherwise return a symbol unique to the given value.
+
+        Args:
+            value: The sympy.Symbol, sympy expression, name, or float to resolve
+                to a unique symbol or float.
+
+        Returns:
+            The unique symbol or value of the parameter as resolved by this
+            resolver.
+        """
+        if isinstance(value, (int, float)):
+            return value
+        if isinstance(value, str):
+            value = sympy.Symbol(value)
+        out = self.param_dict.get(value, None)
+        if out is not None:
+            return out
+        # Create a new symbol
+        symbol = self._next_symbol(value)
+        self.param_dict[value] = symbol
+        self._taken_symbols.add(symbol)
+        return symbol
+
+    # Default object truth, equality, and hash
+    __eq__ = object.__eq__
+    __ne__ = object.__ne__
+    __hash__ = object.__hash__
+
+    def __bool__(self):
+        return True
+
+    def __repr__(self):
+        if self.get_param_name == self.default_get_param_name:
+            return '_ParamFlattener({!r})'.format(self.param_dict)
+        else:
+            return ('_ParamFlattener({!r}, get_param_name={!r})'.format(
+                self.param_dict, self.get_param_name))
+
+    def flatten(self, val: Any) -> Any:
+        """Returns a copy of `val` with any symbols or expressions replaced with
+        new symbols.  `val` can be a `Circuit`, `Gate`, `Operation`, or other
+        type.
+
+        This method mutates the `_ParamFlattener` by storing any new mappings
+        from expression to symbol that is uses on val.  Use `transform_sweep` or
+        `transform_params` after `flatten`.
+
+        Args:
+            val: The value to copy with substituted parameters.
+        """
+        return protocols.resolve_parameters(val, self)
+
+
+class ExpressionMap(dict):
+    """A dictionary where the keys are sympy expressions and symbols and the
```

Comment:
Add comma after the second "and" or split it into two sentences.

Reply:
Done.